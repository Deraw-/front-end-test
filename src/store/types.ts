// tslint:disable-next-line:no-empty-interface
export interface RootState {}

export interface Product extends Object {
	id: number;
	quantity: number;
}

export interface CartProduct extends Product {
	name: string;
	price: number;
	image: string;
}

// This is used to properly type components using mappers from Vuex
// see https://github.com/vuejs/vuex/pull/1121
// tslint:disable-next-line: no-empty-interface
export default interface VuexBindings {
	cartProducts: CartProduct[];
}
