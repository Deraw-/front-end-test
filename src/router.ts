import Vue from 'vue';
import Router from 'vue-router';
import Meta from 'vue-meta';

import i18n, { loadLanguageAsync } from '@/i18n';

// Directly import Anomalies view for faster rendering of first page
import Home from '@/views/Home.vue';

Vue.use(Router);
Vue.use(Meta);

// See https://router.vuejs.org/fr/essentials/navigation.html for help
const router = new Router({
	mode: 'history', // Use the History API, and remove the hash from the URL
	routes: [
		{
			path: '/',
			name: 'home',
			component: Home
		},
		{
			path: '/cart',
			name: 'cart',
			component: () => import(/* webpackChunkName: "cart" */ './views/Cart.vue')
		},
		{
			path: '*', // Everything that doesn't match something
			name: 'notFound',
			component: () => import(/* webpackChunkName: "not-found" */ './views/NotFound.vue')
		}
	],
	// Native-like scroll behavior
	scrollBehavior(to, from, savedPosition) {
		// If there is an hash, this is an anchor, so scroll to it
		if (to.hash) {
			return {
				selector: to.hash
			};
		} else {
			// Else simulate browser behavior by returning to the previous position, or fallback to top
			if (savedPosition) {
				return savedPosition;
			} else {
				return { x: 0, y: 0 };
			}
		}
	}
});

// Resolve translations before each navigation
router.beforeEach((to, from, next) => {
	const lang = i18n.locale;
	loadLanguageAsync(lang).then(() => next());
});

export default router;
