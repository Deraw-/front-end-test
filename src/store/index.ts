import Vue from 'vue';
import Vuex, { StoreOptions } from 'vuex';
import VuexPersistence from 'vuex-persist';
import { RootState } from './types';

// Modules
import cart from './modules/cart';
import products from './modules/products';

// The store is saved in the browser's session
const vuexLocal = new VuexPersistence({
	storage: window.sessionStorage
});

Vue.use(Vuex);

// See https://vuex.vuejs.org/fr/getting-started.html for help
const store: StoreOptions<RootState> = {
	strict: true,
	state: {},
	// See https://vuex.vuejs.org/guide/modules.html for more info on modules
	modules: {
		cart,
		products
	},
	plugins: [
		vuexLocal.plugin
	]
};

export default new Vuex.Store<RootState>(store);
