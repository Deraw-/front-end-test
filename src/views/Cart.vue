<template>
	<section class="page">
		<h2 class="title">
			Your Cart
		</h2>

		<!-- Empty cart -->
		<template v-if="!cartProducts.length">
			<p class="description">
				Your cart is empty!
			</p>

			<RouterLink
				class="btn primary"
				:to="{ name: 'home' }"
			>
				Browse products
			</RouterLink>
		</template>

		<template v-else>
			<CartList :products="cartProducts" />

			<div class="total">
				Total ({{ cartProducts.length }} items): <span class="price">{{ total | euros }}</span>
			</div>

			<div class="suggestion">
				<h2 class="title">
					You might also likeâ€¦
				</h2>

				<ul
					v-if="suggestion"
					class="grid-list"
				>
					<Product
						v-for="product in suggestion"
						:key="product.id"
						:product="product"
					/>
				</ul>
			</div>
		</template>
	</section>
</template>

<script lang="ts">
	import Vue, { VueConstructor } from 'vue';
	import { mapGetters, mapState } from 'vuex';
	import VuexBindings, { CartProduct, ProductType } from '@/store/types';

	import CartList from '@/components/CartList.vue';
	import Product from '@/components/Product.vue';

	export default (Vue as VueConstructor<Vue & VuexBindings>).extend({
		name: 'Cart',
		components: {
			CartList,
			Product
		},
		computed: {
			...mapState('products', ['all']),
			...mapState('cart', ['added']),
			...mapGetters('cart', ['cartProducts']),
			total(): number {
				return this.cartProducts.reduce((total: number, p: CartProduct) => {
					return total + p.price * p.quantity;
				}, 0);
			},
			suggestion() {
				// Filter to not suggest products already in cart
				const suggestion: ProductType[] = this.all.filter((obj: ProductType) => {
					const added = this.added.find((addedObj: ProductType) => obj.id === addedObj.id);

					return !added;
				});

				// Only 4 suggestions
				return suggestion.slice(0, 4);
			}
		}
	});
</script>

<style lang="scss" scoped>
	.total {
		margin-top: 15px;
		margin-right: 20px;

		font-weight: 700;
		text-align: right;
	}

	.price {
		color: #008393;
	}

	.suggestion {
		margin-top: auto;
	}

	.grid-list {
		margin-top: 25px;
	}
</style>
