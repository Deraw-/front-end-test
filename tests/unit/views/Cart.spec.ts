import { shallowMount } from '@vue/test-utils';
import Cart from '@/views/Cart.vue';

import localVue from 'tests-unit/index';
import router from 'tests-unit/router';
import store from '@/store/';

// Build component before each test
let wrapper: any;
beforeEach(() => {
	wrapper = shallowMount(Cart, {
		localVue,
		router,
		store
	});
});

// Tests
describe('Cart.vue', () => {
	it('renders correctly', () => {
		expect(wrapper).toMatchSnapshot();
	});
});
