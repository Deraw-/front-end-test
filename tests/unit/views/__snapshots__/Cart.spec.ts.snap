// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart.vue renders correctly 1`] = `
<section class="page">
  <h2 class="title">
    views.cart.title
  </h2>
  <p class="description">
    views.cart.empty.title
  </p>
  <routerlink-stub to="[object Object]" tag="a" event="click" class="btn primary">
    views.cart.empty.homeLink
  </routerlink-stub>
</section>
`;
